bin/chef-cluster

```diff
@@ -0,0 +1,115 @@
+#!/bin/sh
+#
+# Copyright 2017, SUSE
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+cmd=$1
+workercount=${2:-auto}
+vhostconfig=/etc/apache2/vhosts.d/chef-cluster.conf
+serverconfig=/etc/sysconfig/chef-server
+webroot=/srv/www/chef-cluster
+
+baseport=4001
+
+balancerstart=BEGIN-BALANCER-MEMBER
+balancerend=END-BALANCER-MEMBER
+
+usage() {
+  echo "usage: chef-cluster <enable|disable> [<workercount>|auto]"
+  exit 1
+}
+
+case $cmd in
+  enable)
+    [ "$workercount" == "auto" ] && workercount=$(grep -c ^processor /proc/cpuinfo)
+    [ "$workercount" -ge 2 ] || { echo "ERROR: workercount must be >= 2"; exit 2; }
+
+    # apache configuration
+    mkdir -p "$webroot"
+    a2enmod -q lbmethod_byrequests || a2enmod lbmethod_byrequests
+    test -e "$vhostconfig" || cat <<EOF > $vhostconfig
+# configuration generated by: chef-cluster script
+Listen 4000
+
+<VirtualHost *:4000>
+  DocumentRoot /srv/www/chef-cluster
+  ErrorLog /var/log/apache2/chef-cluster-error.log
+  CustomLog /var/log/apache2/chef-cluster-access.log combined
+
+  <Location />
+    Require all granted
+  </Location>
+
+  <Proxy balancer://chefworkers>
+    # $balancerstart
+    # $balancerend
+  </Proxy>
+
+  <Location /balancer-manager>
+    SetHandler balancer-manager
+    Require host localhost
+    Require host 127.0.0.1
+  </Location>
+
+  RewriteEngine On
+  RewriteCond %{REQUEST_URI} !=/balancer-manager
+  RewriteCond %{DOCUMENT_ROOT}/%{REQUEST_FILENAME} !-f
+  RewriteRule ^/(.*)$ balancer://chefworkers%{REQUEST_URI} [P,QSA,L]
+</VirtualHost>
+EOF
+
+    awk -i inplace "
+      /$balancerstart/ { print; f=1} !f;
+      /$balancerend/ {
+        for (i=$baseport;i<$baseport+$workercount;i++) {
+          print \"    BalancerMember http://127.0.0.1:\"i
+        };
+        print;
+        f=0
+      }
+    " "$vhostconfig"
+    echo "$vhostconfig created/updated"
+
+    # chef-server configuration
+    test -e "$serverconfig" && mv "$serverconfig" "$serverconfig.old"
+    cat <<EOF > $serverconfig
+# configuration generated by: chef-cluster script
+PORT=$baseport
+OPTIONS="-N -c $workercount"
+EOF
```

Comment:
I would actually do something like:

```shell
tmpserverconfig=`mktemp`
cat <<EOF > "$tmpserverconfig"
...
EOF
if cmp -s "$serverconfig" "$tmpserverconfig"; then
  rm -f "$tmpserverconfig"
else
  test -e "$serverconfig" && mv "$serverconfig" "$serverconfig.old"
  mv "$tmpserverconfig" "$serverconfig"
  systemctl restart chef-server
fi
```

This avoid backing up the same config file and restarting chef-server with no reason.

Btw, I've removed the `echo` because I would only keep one (see next comment).

Reply:
Oh, great idea. I'll add this.